fuzz/fuzz_targets/differential-new.rs

```diff
@@ -0,0 +1,85 @@
+#![no_main]
+
+use libfuzzer_sys::arbitrary::{Result, Unstructured};
+use libfuzzer_sys::fuzz_target;
+use wasmtime_fuzzing::generators::{DiffValue, SingleInstModule};
+use wasmtime_fuzzing::oracles::engine::{get_exported_function_signatures, DiffEngine};
+use wasmtime_fuzzing::oracles::{diff_wasmtime, differential, engine};
+
+const NUM_INVOCATIONS: usize = 5;
+
+fuzz_target!(|data: [u8; 1024]| {
+    // errors in `run` have to do with not enough input in `data`, which we
+    // ignore here since it doesn't affect how we'd like to fuzz.
+    drop(run(&data));
+});
+
+fn run(data: &[u8]) -> anyhow::Result<()> {
+    let mut u = Unstructured::new(data);
+
+    // Generate the Wasm module. TODO eventually, this should pick between the
+    // single-instruction and wasm-smith modules, but currently the wasm-smith
+    // module generation will eat up all of the random data, leaving none for
+    // the remaining decisions that follow (e.g., choosing an engine, generating
+    // arguments).
+    let module: &SingleInstModule = u.arbitrary()?;
+    let wasm = module.to_bytes();
+    let features = module.to_features();
+
+    // Choose a right-hand side Wasm engine--this will always be Wasmtime. The
+    // order (execute `lhs` first, then `rhs`) is important because, in some
+    // cases (e.g., OCaml spec interpreter), both sides register signal
+    // handlers; Wasmtime uses these signal handlers for catching various
+    // WebAssembly failures. On certain OSes (e.g. Linux x86_64), the signal
+    // handlers interfere, observable as an uncaught `SIGSEGV`--not even caught
+    // by libFuzzer. By always running Wasmtime second, its signal handlers are
+    // registered most recently and they catch failures appropriately. We create
+    // `rhs` first, however, so we have the option of creating a compatible
+    // Wasmtime engine (e.g., pooling allocator memory differences).
+    let rhs = diff_wasmtime::WasmtimeEngine::arbitrary_with_features(&mut u, &features)
+        .expect("unable to build a Wasmtime `rhs` engine");
+
+    // Choose a left-hand side Wasm engine.
+    let lhs = engine::choose(&mut u, &features, &rhs)
+        .expect("unable to choose an engine to fuzz against");
+
+    // Instantiate each engine and try each exported functions with various
+    // values.
+    let mut lhs_instance = lhs
+        .instantiate(&module.to_bytes())
+        .expect("failed to instantiate `lhs` module");
+    let mut rhs_instance = rhs
+        .instantiate(&module.to_bytes())
+        .expect("failed to instantiate `rhs` module");
+    for (name, signature) in get_exported_function_signatures(&wasm)
+        .expect("failed to extract exported function signatures")
+    {
+        let mut invocations = 0;
+        loop {
```

Comment:
This loop may not play nicely with fuel right now since if the module has wasm-smith-injected fuel the fuel is never reset and otherwise with Wasmtime-injected fuel I'm not sure that's ever reset either.

Reply:
Yeah, we should reset fuel for each call.

Reply:
Right now, with only single-instruction modules, this shouldn't be too much of an issue but once we add in the wasm-smith modules this will be necessary; what is the right way to reset this?

Reply:
Currently there's no way to reset the fuel, we'd have to add an exported function in the `wasm-smith` generated module (or export the global) for it to get modified by the embedder.